File: go-version/leetcode/LC2311/longest_subsequence.go
package LC2311

import (
	"math/bits"
	"strconv"
	"strings"
)

/**
 * @Author: cutejiuge cutejiuge@163.com
 * @Date: 2025/6/26 下午11:05
 * @FilePath: longest_subsequence
 * @Description:
 */

func longestSubsequence(s string, k int) int {
	n, m := len(s), bits.Len(uint(k))
	if n < m {
		// 可以全选
		return n
	}
	ans := m //后缀长度
	sufVal, _ := strconv.ParseInt(s[n-m:], 2, 0)
	if int(sufVal) > k {
		ans--
	}
	return ans + strings.Count(s[:n-m], "0")
}

File: python-version/LC2311/longest_subsequence.py
class Solution:
    @staticmethod
    def longestSubsequence(s: str, k: int) -> int:
        n, m = len(s), k.bit_length()
        if n < m:
            return n
        suf_val = int(s[n-m:], 2)
        ans = m - 1 if suf_val > k else m
        return ans + s[:n-m].count('0')
